### PR TITLE
Page about granting access to projects and namespaces

### LINKED ISSUE #47: How To Grant Access to Namespaces/Projects
We need a how-to which describes how access management in a Namespace works and how it is managed via API (YAML) and OpenShift Webconsole.

This is needed because we have different ways to grant access:
* User
* Organization
* Team
* ServiceAccount

And the OpenShift webconsole doesn't provide a good user experience (Just a form to clicky-bunty a RoleBinding) plus it doesn't have autocomplet for Groups. So the user has to know how e.g. the Group is called for an Organization or a Team.

Links:
* https://docs.openshift.com/container-platform/4.9/authentication/using-rbac.html

### PR DESCRIPTION
**Pull request by akosma**
Resolves #47 .

## Checklist

- [x] Try to isolate changes into separate PRs (to build a better changelog).
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `how-to`, `tutorial`, `reference`, `explanation`, `cicd`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues if applicable.

### CODE REVIEW

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 1]
```diff
@@ -0,0 +1,27 @@
+= Granting Access to Projects and Namespaces
+
+In the xref:how-to/manage-projects-and-namespaces.adoc[Managing Projects and Namespaces] page it's explained how to create projects and namespaces.
+This page explains in detail how to manage users, groups, and teams within those projects, assigning rights and entitlements.
+
+IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/4.9/authentication/using-rbac.html[official OpenShift 4.9 documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.
+
+== Organizations and Teams in {product}
+
+Organizations and teams are reflected as OpenShift 4.9 groups.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:10]:
```suggestion
Organizations and teams are reflected as OpenShift groups.
```

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+For each and every organization, there is a group with the same name; and for each and every team, there is a group with the same name, but prefixed with the name of the parent organization.
+For example: `acme+accounting` would be a valid team name in {product} for the Accounting team within the ACME organization.
+
+IMPORTANT: OpenShift groups don't have any notion of hierarchy! From the point of view of OpenShift, they all belong to the same flat structure.
+
+== Granting Privileges
+
+A `RoleBinding` object grants the creator of a Project or Namespace admin privileges.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:18]: That one isn't clear. Which `RoleBinding`?

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+This is a default behavior in OpenShift.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:19]: That's not true IMHO? Or what does "This" refer to?

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+This means that by default, all members of the owning organization are granted admin privileges by referencing the corresponding OpenShift Group.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:21]: IMHO this is APPUiO Cloud specific?

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+Both RoleBindings can be deleted and replaced by something else.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:23]: Which "both"? How are they called?

**Review by tobru — CHANGES_REQUESTED** [review]
Some clarification needed, maybe some input from Tarazed / @corvus-ch is needed. I would add some examples of RoleBindings (Yaml) to make things more clear. And if possible, add some hints how to get the correct name of the groups (organization and team group), not entirely sure how to find that...

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 6]
```diff
@@ -0,0 +1,27 @@
+= Granting Access to Projects and Namespaces
+
+In the xref:how-to/manage-projects-and-namespaces.adoc[Managing Projects and Namespaces] page it's explained how to create projects and namespaces.
+This page explains in detail how to manage users, groups, and teams within those projects, assigning rights and entitlements.
+
+IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/4.9/authentication/using-rbac.html[official OpenShift 4.9 documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:6]:
```suggestion
IMPORTANT: The contents of this page rely on knowledge provided by the https://docs.openshift.com/container-platform/latest/authentication/using-rbac.html[official OpenShift documentation about Role-Base Access Control (RBAC)], highlighting the differences in {product}.
```

File: docs/modules/ROOT/pages/how-to/grant-access.adoc (hunk 2 above)
Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:18]:
```suggestion
The RoleBinding object named `admin` grants the creator of a Project or Namespace admin privileges.
```

File: docs/modules/ROOT/pages/how-to/grant-access.adoc (hunk 4 above)
Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:21]:
```suggestion
In addition, the RoleBinding named `namespace-owner` grants the group named equal to the Organization owning the namespace admin privileges.
This means that by default, all members of the owning organization are granted admin privileges.
```

File: docs/modules/ROOT/pages/how-to/grant-access.adoc (hunk 5 above)
Comment (corvus-ch) [reply docs/modules/ROOT/pages/how-to/grant-access.adoc:23]: Should be addressed by my suggestions.

Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:23]:
```suggestion
Both RoleBindings can be changed or deleted and replaced by something else.
```

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+A `RoleBinding` object grants the creator of a Project or Namespace admin privileges.
+This is a default behavior in OpenShift.
+
+This means that by default, all members of the owning organization are granted admin privileges by referencing the corresponding OpenShift Group.
+
+Both RoleBindings can be deleted and replaced by something else.
+
+CAUTION: Modifications to RoleBindings can prevent users from accessing their projects; in other words, you might end up locked out of your project!
+
+Access can be granted to any OpenShift Group (This includes that namespace owned by organization A can be managed by members of organization B or only a team of organization B)
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:27]:
```suggestion
Access can be granted to any OpenShift Group.
This includes that namespace owned by organization A can be managed by members of organization B or only a team of organization B.
```

**Review by corvus-ch — CHANGES_REQUESTED** [review]
While it is no guarantee, using https://docs.openshift.com/container-platform/latest instead of the version specific URLs is less likely to break.

I also added suggestion with RoleBindings in question.

File: docs/modules/ROOT/pages/how-to/grant-access.adoc [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+IMPORTANT: OpenShift groups don't have any notion of hierarchy! From the point of view of OpenShift, they all belong to the same flat structure.
+
+== Granting Privileges
+
+The RoleBinding object named `admin` grants the creator of a Project or Namespace admin privileges.
+This is a default behavior in OpenShift.
+
+In addition, the RoleBinding named `namespace-owner` grants the group named equal to the Organization owning the namespace admin privileges.
+This means that by default, all members of the owning organization are granted admin privileges.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-to/grant-access.adoc:22]: These paragraphs are not correct.

1. We patch the `admin` RoleBinding to grant the organization admin privileges on the project, cf. https://hub.syn.tools/appuio-cloud/references/policies/10_generate_default_rolebinding_in_ns.html
2. We create the `namespace-owner` RoleBinding, which allows the organization to view, edit and delete the Namespace object itself. Edit restrictions apply, cf. https://hub.syn.tools/appuio-cloud/references/policies/02_validate_namespace_metadata.html

Maybe rewrite the paragraphs along the lines of the suggestion below (I'm not happy  with the wording, feel free to improve it).

```suggestion
On {product}, the RoleBinding object named `admin` grants the group named equal to the Organization owning the Project or Namespace admin privileges.
This differs from the default behavior in OpenShift, where the default `admin` RoleBinding only grants the creator of a Project admin privileges.

In addition, the RoleBinding named `namespace-owner` gives the group named equal to the Organization of the Namespace object permissions to view, edit and delete the object.
```

**Review by simu — CHANGES_REQUESTED** [review]

**Review by simu — APPROVED** [review]

**Comment by tobru** [comment]
@corvus-ch Do we need to add some example RoleBindings which make use of the organization or team groups? And do you know how one could look-up the exact name of the organization or team group? I think we should describe that...

**Comment by corvus-ch** [comment]
> @corvus-ch Do we need to add some example RoleBindings which make use of the organization or team groups?

I was thinking the same, but then I failed to come up with a good example. Preferably, the example would reflect a real world use case.

> And do you know how one could look-up the exact name of the organization or team group? I think we should describe that...

No, there is no "look-up". We might add that to https://github.com/appuio/cloud-portal (copy role name on both Organizations and Teams).

**Comment by tobru** [comment]
> I was thinking the same, but then I failed to come up with a good example. Preferably, the example would reflect a real world use case.

What about: Create two Teams, one "view only", one "admin" and create matching RoleBindings, then remove access on the organization level?

> No, there is no "look-up". We might add that to https://github.com/appuio/cloud-portal (copy role name on both Organizations and Teams).

Isn't the name of the organization group the same as the organization name? So we could refer to that. And for the team it's the organization name plus the team name, so we could say: Hey, look in the Portal for the exact names. Another idea could be to look into the generated RoleBindings for the organization group name. Maybe `oc get groups` only returns the group you have access to? Not sure though... Or potentially there is another command to get group membership? I'm also not sure about that.

**Comment by corvus-ch** [comment]
> What about: Create two Teams, one "view only", one "admin" and create matching RoleBindings, then remove access on the organization level?

Yes, that was also my morning shower thought. We might go even further and give an example for the three levels that were available on APPUiO Public.

**Comment by corvus-ch** [comment]
> Isn't the name of the organization group the same as the organization name? So we could refer to that.

Correct

> And for the team it's the organization name plus the team name

Correct.

But there is no "look-up" yet, at least to my knowledge.

> Maybe `oc get groups` only returns the group you have access to? Not sure though... Or potentially there is another command to get group membership? I'm also not sure about that.

That would be nice, but as far as I am aware, only cluster-viewer can list groups. I might be wrong, though.

> so we could say: Hey, look in the Portal for the exact names.

That was my suggestion, yes. But not only, hey go look there, but give them a "copy group name" button.

**Comment by tobru** [comment]
@corvus-ch Can you bring this docs over the finishing line with @akosma so that we can merge it?

**Review by tobru — APPROVED** [review]

**Review by corvus-ch — APPROVED** [review]
I made some last clarifications. Once they are in, this is good to go. We can add examples at any time in the future. For now, it is more important to have this published.

**Pull request merged**